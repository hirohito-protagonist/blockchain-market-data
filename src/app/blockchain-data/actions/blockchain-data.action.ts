import { Action } from '@ngrx/store';

export enum ActionTypes {
  FetchData = '[Blockchain data] fetch data'
}

export class FetchData implements Action {
  readonly type = ActionTypes.FetchData;

  constructor(public payload: { key: string; query: any; }) {}
}


export type ActionType = FetchData;
