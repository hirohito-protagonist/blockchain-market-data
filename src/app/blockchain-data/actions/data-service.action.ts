import { Action } from '@ngrx/store';

export enum DataServiceType {
  Ticker = 'ticker',
  ToBTC = 'tobtc'
}

export enum ActionTypes {
  Request = '[Blockchain data] data service request',
  Response = '[Blockchain data] data service response'
}

export class Request implements Action {
  readonly type = ActionTypes.Request;

  constructor(public payload: { key: DataServiceType; query: any; }) {}
}

export class Response implements Action {
  readonly type = ActionTypes.Response;

  constructor(public payload: { key: DataServiceType; response: any; }) {}
}

export type ActionType = Request | Response;
