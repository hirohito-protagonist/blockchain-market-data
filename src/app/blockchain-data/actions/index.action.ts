import * as fromDataServiceAction from './data-service.action';
import * as fromBlockchainDataAction from './blockchain-data.action';

export {
  fromDataServiceAction,
  fromBlockchainDataAction
};
