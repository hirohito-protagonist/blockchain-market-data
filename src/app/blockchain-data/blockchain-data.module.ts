import { NgModule, ModuleWithProviders } from '@angular/core';

import { StoreModule } from '@ngrx/store';

import { reducers } from './reducers/index.reducer';

@NgModule({
  imports: [
    StoreModule.forFeature('blockchain-data', reducers)
  ]
})
export class BlockchainDataModule {

  static forRoot(): ModuleWithProviders {

    return {
      ngModule: BlockchainDataModule,
      providers: []
    };
  }
}
