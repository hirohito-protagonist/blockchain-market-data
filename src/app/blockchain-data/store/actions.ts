import { createAction, props } from '@ngrx/store';
import { DataServiceType, QueryType, DataResponseType } from './../blockchain-data.type';

export const fetchData = createAction(
  '[Blockchain Data Feature API] Fetch data',
  props<{ key: DataServiceType; query: QueryType; }>()
);

export const request = createAction(
  '[Blockchain Data Feature API] Data service request',
  props<{ key: DataServiceType; query: QueryType; }>()
);

export const response = createAction(
  '[Blockchain Data Feature API] Data service response',
  props<{ key: DataServiceType; response: DataResponseType; }>()
);
