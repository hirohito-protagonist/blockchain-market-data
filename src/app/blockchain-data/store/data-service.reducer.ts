import { createReducer, on } from '@ngrx/store';
import { request, response } from './../store/actions';
import { DataServiceType, QueryType, DataResponseType } from './../blockchain-data.type';

interface DataServiceNode<R, Q> {
  isFetching: boolean;
  response: R;
  lastUpdate: number;
  query: Q;
}

export interface DataServiceState {
  [key: string]: DataServiceNode<DataResponseType, QueryType>;
}

const initialState: DataServiceState = {
  [DataServiceType.Ticker]: {
    isFetching: false,
    response: null,
    lastUpdate: 0,
    query: null
  },
  [DataServiceType.ToBTC]: {
    isFetching: false,
    response: null,
    lastUpdate: 0,
    query: null
  },
  [DataServiceType.Stats]: {
    isFetching: false,
    response: null,
    lastUpdate: 0,
    query: null
  },
  [DataServiceType.Charts]: {
    isFetching: false,
    response: null,
    lastUpdate: 0,
    query: null
  }
};

export const reducer = createReducer(
  initialState,
  on(
    request,
    (state, { key, query }) => {

      if (state[key]) {

        return {
          ...state,
          [key]: {
            ...state[key],
            isFetching: true,
            response: null,
            query
          }
        };
      }
      return state;
    }
  ),
  on(
    response,
    (state, { key, response }) => {

      if (state[key]) {

        return {
          ...state,
          [key]: {
            ...state[key],
            isFetching: false,
            response,
            lastUpdate: Date.now()
          }
        };
      }
      return state;
    }
  )
);
