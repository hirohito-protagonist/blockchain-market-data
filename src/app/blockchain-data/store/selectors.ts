import { createFeatureSelector, createSelector } from '@ngrx/store';
import { DataServiceType } from './../blockchain-data.type';
import { DataServiceState } from './store';
import * as fromDataService from './data-service.reducer';

const getFeatureRootState = createFeatureSelector('blockchain-data');
const getDataServiceState = createSelector(getFeatureRootState, (s: DataServiceState) => s.data);
export const getServiceDataNode = (dataServiceType: DataServiceType) =>
  createSelector(getDataServiceState, (s: fromDataService.DataServiceState) => s[dataServiceType]);

export const lastUpdate = (dataServiceType: DataServiceType) =>
  createSelector(getServiceDataNode(dataServiceType), (node) => node.lastUpdate);

export const response = (dataServiceType: DataServiceType) =>
  createSelector(getServiceDataNode(dataServiceType), (node) => node.response);
