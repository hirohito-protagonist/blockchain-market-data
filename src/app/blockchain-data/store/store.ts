import { StoreModule } from '@ngrx/store';
import { EffectsModule } from '@ngrx/effects';
import * as fromDataService from './data-service.reducer';
import { DataServiceEffects } from './effects';

export interface DataServiceState {
  data: fromDataService.DataServiceState;
  version: string;
}

export function featureVersion() {
  return '1.0.0';
}

export const reducers = {
  data: fromDataService.reducer,
  version: featureVersion
};


export const Store = [
  StoreModule.forFeature('blockchain-data', reducers),
  EffectsModule.forFeature([ DataServiceEffects ])
];

