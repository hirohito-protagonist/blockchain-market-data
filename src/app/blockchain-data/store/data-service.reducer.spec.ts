import { reducer } from './data-service.reducer';
import { DataServiceType } from './../blockchain-data.type';
import { request, response } from './../store/actions';

describe('Data service reducer', () => {

  describe('init', () => {

    it('should return default state', () => {

      // Given
      const action = {} as any;

      // When
      const result = reducer(void(0), action);

      // Then
      expect(result).toMatchSnapshot();
    });
  });

  describe('[Blockchain data] data service request action', () => {

    [
      {
        serviceType: DataServiceType.ToBTC,
        query: { currency: 'USD', value: 200 }
      },
      {
        serviceType: DataServiceType.Stats,
        query: null
      },
      {
        serviceType: DataServiceType.Ticker,
        query: null
      }
    ].forEach((testData) => {

      it(`should set isFetching as true and update query for ${testData.serviceType}`, () => {

        // Given
        const action = request({
          key: testData.serviceType, query: testData.query
        });

        // When
        const result = reducer(void(0), action);

        // Then
        expect(result).toMatchSnapshot();
      });
    });


    it('should not update state for undefined node', () => {

      // Given
      const action = request({
        key: '' as any, query: { currency: 'USD', value: 200 }
      });

      // When
      const result = reducer(void(0), action);

      // Then
      expect(result).toMatchSnapshot();
    });
  });

  describe('[Blockchain data] data service response action', () => {

    [
      {
        serviceType: DataServiceType.ToBTC,
        query: { currency: 'USD', value: 200 },
        response: 123
      },
      {
        serviceType: DataServiceType.Stats,
        query: null,
        response: {}
      },
      {
        serviceType: DataServiceType.Ticker,
        query: null,
        response: {}
      }
    ].forEach((testData) => {

      it(`should set isFetching as false and update response for ${testData.serviceType}`, () => {

        spyOn(Date, 'now').and.returnValue(123);

        // Given
        const requestAction = request({
          key: testData.serviceType, query: testData.query
        });
        const responseAction = response({
          key: testData.serviceType, response: testData.response
        });

        // When
        const requestResult = reducer(void(0), requestAction);
        const responseResult = reducer(requestResult, responseAction);

        // Then
        expect(responseResult).toMatchSnapshot();
      });
    });


    it('should not update state for undefined node', () => {

      // Given
      const requestAction = request({
        key: DataServiceType.ToBTC, query: { currency: 'USD', value: 200 }
      });
      const responseAction = response({
        key: '' as any, response: 123
      });

      // When
      const requestResult = reducer(void(0), requestAction);
      const responseResult = reducer(requestResult, responseAction);

      // Then
      expect(responseResult).toMatchSnapshot();
    });
  });

});
