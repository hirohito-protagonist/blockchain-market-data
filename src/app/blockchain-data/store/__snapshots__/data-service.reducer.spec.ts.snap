// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Data service reducer [Blockchain data] data service request action should not update state for undefined node 1`] = `
Object {
  "charts": Object {
    "isFetching": false,
    "lastUpdate": 0,
    "query": null,
    "response": null,
  },
  "stats": Object {
    "isFetching": false,
    "lastUpdate": 0,
    "query": null,
    "response": null,
  },
  "ticker": Object {
    "isFetching": false,
    "lastUpdate": 0,
    "query": null,
    "response": null,
  },
  "tobtc": Object {
    "isFetching": false,
    "lastUpdate": 0,
    "query": null,
    "response": null,
  },
}
`;

exports[`Data service reducer [Blockchain data] data service request action should set isFetching as true and update query for stats 1`] = `
Object {
  "charts": Object {
    "isFetching": false,
    "lastUpdate": 0,
    "query": null,
    "response": null,
  },
  "stats": Object {
    "isFetching": true,
    "lastUpdate": 0,
    "query": null,
    "response": null,
  },
  "ticker": Object {
    "isFetching": false,
    "lastUpdate": 0,
    "query": null,
    "response": null,
  },
  "tobtc": Object {
    "isFetching": false,
    "lastUpdate": 0,
    "query": null,
    "response": null,
  },
}
`;

exports[`Data service reducer [Blockchain data] data service request action should set isFetching as true and update query for ticker 1`] = `
Object {
  "charts": Object {
    "isFetching": false,
    "lastUpdate": 0,
    "query": null,
    "response": null,
  },
  "stats": Object {
    "isFetching": false,
    "lastUpdate": 0,
    "query": null,
    "response": null,
  },
  "ticker": Object {
    "isFetching": true,
    "lastUpdate": 0,
    "query": null,
    "response": null,
  },
  "tobtc": Object {
    "isFetching": false,
    "lastUpdate": 0,
    "query": null,
    "response": null,
  },
}
`;

exports[`Data service reducer [Blockchain data] data service request action should set isFetching as true and update query for tobtc 1`] = `
Object {
  "charts": Object {
    "isFetching": false,
    "lastUpdate": 0,
    "query": null,
    "response": null,
  },
  "stats": Object {
    "isFetching": false,
    "lastUpdate": 0,
    "query": null,
    "response": null,
  },
  "ticker": Object {
    "isFetching": false,
    "lastUpdate": 0,
    "query": null,
    "response": null,
  },
  "tobtc": Object {
    "isFetching": true,
    "lastUpdate": 0,
    "query": Object {
      "currency": "USD",
      "value": 200,
    },
    "response": null,
  },
}
`;

exports[`Data service reducer [Blockchain data] data service response action should not update state for undefined node 1`] = `
Object {
  "charts": Object {
    "isFetching": false,
    "lastUpdate": 0,
    "query": null,
    "response": null,
  },
  "stats": Object {
    "isFetching": false,
    "lastUpdate": 0,
    "query": null,
    "response": null,
  },
  "ticker": Object {
    "isFetching": false,
    "lastUpdate": 0,
    "query": null,
    "response": null,
  },
  "tobtc": Object {
    "isFetching": true,
    "lastUpdate": 0,
    "query": Object {
      "currency": "USD",
      "value": 200,
    },
    "response": null,
  },
}
`;

exports[`Data service reducer [Blockchain data] data service response action should set isFetching as false and update response for stats 1`] = `
Object {
  "charts": Object {
    "isFetching": false,
    "lastUpdate": 0,
    "query": null,
    "response": null,
  },
  "stats": Object {
    "isFetching": false,
    "lastUpdate": 123,
    "query": null,
    "response": Object {},
  },
  "ticker": Object {
    "isFetching": false,
    "lastUpdate": 0,
    "query": null,
    "response": null,
  },
  "tobtc": Object {
    "isFetching": false,
    "lastUpdate": 0,
    "query": null,
    "response": null,
  },
}
`;

exports[`Data service reducer [Blockchain data] data service response action should set isFetching as false and update response for ticker 1`] = `
Object {
  "charts": Object {
    "isFetching": false,
    "lastUpdate": 0,
    "query": null,
    "response": null,
  },
  "stats": Object {
    "isFetching": false,
    "lastUpdate": 0,
    "query": null,
    "response": null,
  },
  "ticker": Object {
    "isFetching": false,
    "lastUpdate": 123,
    "query": null,
    "response": Object {},
  },
  "tobtc": Object {
    "isFetching": false,
    "lastUpdate": 0,
    "query": null,
    "response": null,
  },
}
`;

exports[`Data service reducer [Blockchain data] data service response action should set isFetching as false and update response for tobtc 1`] = `
Object {
  "charts": Object {
    "isFetching": false,
    "lastUpdate": 0,
    "query": null,
    "response": null,
  },
  "stats": Object {
    "isFetching": false,
    "lastUpdate": 0,
    "query": null,
    "response": null,
  },
  "ticker": Object {
    "isFetching": false,
    "lastUpdate": 0,
    "query": null,
    "response": null,
  },
  "tobtc": Object {
    "isFetching": false,
    "lastUpdate": 123,
    "query": Object {
      "currency": "USD",
      "value": 200,
    },
    "response": 123,
  },
}
`;

exports[`Data service reducer init should return default state 1`] = `
Object {
  "charts": Object {
    "isFetching": false,
    "lastUpdate": 0,
    "query": null,
    "response": null,
  },
  "stats": Object {
    "isFetching": false,
    "lastUpdate": 0,
    "query": null,
    "response": null,
  },
  "ticker": Object {
    "isFetching": false,
    "lastUpdate": 0,
    "query": null,
    "response": null,
  },
  "tobtc": Object {
    "isFetching": false,
    "lastUpdate": 0,
    "query": null,
    "response": null,
  },
}
`;
